File: csharp/src/DataStructures/Queue.cs
namespace DSA.DataStructures;

/// <summary>
/// Queue implementation using a linked list.
/// Time: O(1) for all operations
/// Space: O(n)
/// </summary>
public class Queue<T>
{
    private class Node
    {
        public T Value { get; }
        public Node? Next { get; set; }

        public Node(T value)
        {
            Value = value;
        }
    }

    private Node? _head;
    private Node? _tail;
    private int _size;

    public int Count => _size;
    public bool IsEmpty => _size == 0;

    public void Enqueue(T value)
    {
        var node = new Node(value);
        if (_tail != null)
        {
            _tail.Next = node;
        }
        else
        {
            _head = node;
        }
        _tail = node;
        _size++;
    }

    public T? Dequeue()
    {
        if (_head == null)
            return default;

        var value = _head.Value;
        _head = _head.Next;
        if (_head == null)
            _tail = null;
        _size--;
        return value;
    }

    public T? Peek()
    {
        return _head == null ? default : _head.Value;
    }

    public void Clear()
    {
        _head = null;
        _tail = null;
        _size = 0;
    }
}

File: php/src/DataStructures/Queue.php
<?php

declare(strict_types=1);

namespace DSA\DataStructures;

/**
 * Queue implementation using a linked list.
 * Time: O(1) for all operations
 * Space: O(n)
 *
 * @template T
 */
class Queue
{
    private ?QueueNode $head = null;
    private ?QueueNode $tail = null;
    private int $size = 0;

    public function enqueue(mixed $value): void
    {
        $node = new QueueNode($value);
        if ($this->tail !== null) {
            $this->tail->next = $node;
        } else {
            $this->head = $node;
        }
        $this->tail = $node;
        $this->size++;
    }

    public function dequeue(): mixed
    {
        if ($this->head === null) {
            return null;
        }
        $value = $this->head->value;
        $this->head = $this->head->next;
        if ($this->head === null) {
            $this->tail = null;
        }
        $this->size--;
        return $value;
    }

    public function peek(): mixed
    {
        return $this->head?->value;
    }

    public function isEmpty(): bool
    {
        return $this->size === 0;
    }

    public function size(): int
    {
        return $this->size;
    }

    public function clear(): void
    {
        $this->head = null;
        $this->tail = null;
        $this->size = 0;
    }
}

/**
 * @internal
 */
class QueueNode
{
    public function __construct(
        public mixed $value,
        public ?QueueNode $next = null
    ) {}
}
